File: contests_atcoder/abc175/abc175_e.cpp
#include <iostream>
using namespace std;

int r, c, k;
int items[3000][3000];
long dp[3000][3000][4];

main() {
    cin >> r >> c >> k;
    
    for(int i = 0; i < k; i++){
        int y, x, v;
        cin >> y >> x >> v;
        y--, x--;
        items[y][x] = v;
    }

    for(int y = 0; y < r; y++){
        for(int x = 0; x < c; x++){
            if(y != 0){
                dp[y][x][0] = max(max(dp[y - 1][x][0], dp[y - 1][x][1]), max(dp[y - 1][x][2], dp[y - 1][x][3]));
                dp[y][x][1] = dp[y][x][0] + items[y][x];
            }

            if(x != 0){
                dp[y][x][0] = max(dp[y][x][0], dp[y][x - 1][0]);
                for(int i = 0; i < 3; i++){
                    dp[y][x][i + 1] = max(dp[y][x][i + 1], dp[y][x - 1][i + 1]);
                    dp[y][x][i + 1] = max(dp[y][x][i + 1], dp[y][x - 1][i] + items[y][x]);
                }
            }
            else{
                dp[y][x][1] = dp[y][x][0] + items[y][x];
            }
        }
    }

    cout << max(max(dp[r - 1][c - 1][0], dp[r - 1][c - 1][1]), max(dp[r - 1][c - 1][2], dp[r - 1][c - 1][3])) << endl;
}

File: contests_atcoder/abc175/abc175_e.py
r, c, k = map(int, input().split())
items = [[0] * c for _ in range(r)]

for _ in range(k):
    y, x, v = map(int, input().split())
    items[y-1][x-1] = v

dp0 = [[0] * c for __ in range(r)]
dp1 = [[0] * c for __ in range(r)]
dp2 = [[0] * c for __ in range(r)]
dp3 = [[0] * c for __ in range(r)]

for y in range(r):
    for x in range(c):
        if y != 0:
            dp0[y][x] = max(dp0[y - 1][x], dp1[y - 1][x], dp2[y - 1][x], dp3[y - 1][x])
            dp1[y][x] = max(dp0[y - 1][x], dp1[y - 1][x], dp2[y - 1][x], dp3[y - 1][x]) + items[y][x]
        
        if x != 0:
            dp0[y][x] = max(dp0[y][x], dp0[y][x - 1])
            dp1[y][x] = max(dp0[y][x - 1] + items[y][x], dp1[y][x - 1], dp1[y][x])
            dp2[y][x] = max(dp1[y][x - 1] + items[y][x], dp2[y][x - 1], dp2[y][x])
            dp3[y][x] = max(dp2[y][x - 1] + items[y][x], dp3[y][x - 1], dp3[y][x])
        else:
            dp1[y][x] = dp0[y][x] + items[y][x]

print(max(dp0[-1][-1], dp1[-1][-1], dp2[-1][-1], dp3[-1][-1]))